### PR TITLE
[Refactoring] Fix code issues panel and custom reporting

### LINKED ISSUE #3576: Roslyn code issues are not configurable anymore
`AnalyzersFromAssembly` still exists, mirroring functionality of `MonoDevelopWorkspaceDiagnosticAnalyzerProviderService`.

The `AnalysisOptionsPanel` only gets analyzers from `AnalyzersFromAssembly`, which doesn't have `Microsoft.CodeAnalysis.CSharp`. So we don't bridge UI for that in MD.

The other problem is that we don't pass in specific diagnostic options on compilation requests:
http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/CompilationOptions.cs,198

### PR DESCRIPTION
**Pull request by Therzok**
Fixes #3576

### CODE REVIEW

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs [hunk 1]
```diff
@@ -49,13 +49,49 @@ namespace MonoDevelop.CodeIssues
...
+
+			try {
+				var language = CodeRefactoringService.MimeTypeToLanguage (analysisDocument.Editor.MimeType);
+				var alreadyAdded = new HashSet<Type> ();
+
+				var table = new Dictionary<string, CodeDiagnosticDescriptor> ();
+
+				diagnostics = await CodeRefactoringService.GetCodeDiagnosticsAsync (analysisDocument.DocumentContext, language, cancellationToken);
+				foreach (var diagnostic in diagnostics) {
+					if (alreadyAdded.Contains (diagnostic.DiagnosticAnalyzerType))
```
Comment (KirillOsenkov) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs]: >alreadyAdded.Contains [](start = 9, length = 21)

Just use if (!alreadyAdded.Add()) continue; // Add on HashSet<T> returns false if it was already present

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs [hunk 2]
```diff
@@ -187,10 +214,28 @@ static bool SkipContext (DocumentContext ctx)
...
 
+		static async Task<Diagnostic> ToDiagnosticAsync (this DiagnosticData data, AnalysisDocument analysisDocument, CancellationToken cancellationToken)
+		{
+			var project = analysisDocument.DocumentContext.AnalysisDocument.Project;
+			var location = await data.DataLocation.ConvertLocationAsync (project, cancellationToken).ConfigureAwait (false);
+			var additionalLocations = await data.AdditionalLocations.ConvertLocationsAsync (project, cancellationToken).ConfigureAwait (false);
+
+			var severity = diagnosticTable [data.Id].GetSeverity (data.Id, data.Severity);
+			return Diagnostic.Create (
+				data.Id, data.Category, data.Message, severity, data.DefaultSeverity,
```
Comment (KirillOsenkov) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs]: >data.Id, [](start = 4, length = 8)

What do you think about having each argument on its own line?

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs [hunk 3]
```diff
@@ -49,13 +49,49 @@ namespace MonoDevelop.CodeIssues
...
 		{
 			return (ctx.IsAdHocProject || !(ctx.Project is MonoDevelop.Projects.DotNetProject));
 		}
 
+		static async Task GetDescriptorTable (AnalysisDocument analysisDocument, CancellationToken cancellationToken)
+		{
+			if (diagnosticTable != null)
+				return;
+			
+			bool locked = await diagnosticLock.WaitAsync (Timeout.Infinite, cancellationToken).ConfigureAwait (false);
```
Comment (KirillOsenkov) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs]: >locked [](start = 8, length = 6)

why not just a simple lock?

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs (hunk 1 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs]: This was copy pasted.

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs (hunk 3 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs]: Events fire on task.run, better not lock the whole thread, but actually yield the task if something else can run in the meantime.

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs (hunk 2 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeIssues/CodeDiagnosticRunner.cs]: This was copy pasted from the roslyn repo :p

**Comment by Therzok** [comment]
Merging to master, as this was merged to 15.6

**Comment by Therzok** [comment]
Will fix comments in another PR

**Pull request merged**